ext/storage_sources/s3_store/s3_storage_source.cpp

```diff
@@ -175,120 +146,62 @@ S3SourcePath(const std::string &dir, const std::string &name)
     return (dir + "/" + strippedName);
 }
 
-// Return if the file exists. First checks the cache, and then the S3 Bucket.
+// Return if the file exists, via looking into the S3 Bucket. Also, if the file exists, fetch the
+// file size.
 static int
-S3FileExists(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session, const char *name, bool *fileExists)
+S3FileExistsAndGetSize(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session, const char *name,
+  bool *fileExists, wt_off_t *size)
 {
     int ret = 0;
     S3FileSystem *fs = (S3FileSystem *)fileSystem;
     S3Storage *s3 = FS2S3(fileSystem);
+    size_t objectSize;
 
-    // Check if file exists in the cache.
-    *fileExists = S3CacheExists(fileSystem, name);
-    if (*fileExists) {
-        s3->log->LogDebugMessage("S3FileExists: Found file in cache.");
-        return (ret);
-    }
+    *fileExists = false;
 
-    // It's not in the cache, try the S3 bucket.
-    size_t objectSize;
     s3->statistics.objectExistsCount++;
     if ((ret = fs->connection->ObjectExists(name, *fileExists, objectSize)) != 0)
-        s3->log->LogErrorMessage("S3FileExists: ObjectExists request to S3 failed.");
-    else
-        s3->log->LogDebugMessage("S3FileExists: Found file in S3.");
-
-    return (ret);
-}
+        s3->log->LogErrorMessage("S3FileExistsAndGetSize: ObjectExists request to S3 failed.");
+    else {
+        if (fileExists)
+            s3->log->LogDebugMessage("S3FileExistsAndGetSize: Found file in S3.");
+        else
+            s3->log->LogDebugMessage("S3FileExistsAndGetSize: File not found in S3.");
+    }
 
-// Checks whether the given file exists in the cache.
-static bool
-S3CacheExists(WT_FILE_SYSTEM *fileSystem, const std::string &name)
-{
-    const std::string path = S3CachePath(((S3FileSystem *)fileSystem)->cacheDir, name);
-    return (LocalFileExists(path));
-}
+    if (*fileExists && size != NULL)
+        *size = objectSize;
 
-// Checks whether a file corresponding to the provided path exists locally.
-static bool
-LocalFileExists(const std::string &path)
-{
-    std::ifstream f(path);
-    return (f.good());
+    return (ret);
```

Comment:
This works in kind of a strange way - it seems we can return non-zero if the file doesn't exist. But it's not an error for the file not to exist (since we presumably don't know, and that's why we're checking.)

What about having this return 0 in all non-error cases, and forcing the caller to rely on `*found`?

If we just make this `S3FileGetSize`, then we can return "not found" as an error, and `S3FileExists` can look specifically for that "not found". (And return any other error.)

Reply:
Agreed, the semantics of this function could be cleaned further.

Reply:
Thanks, it reads a lot more clearly now!